### ISSUE TITLE
Add support for `--workspaces` to run on all npm workspaces

### ISSUE THREAD

**Issue by voxpelli** [bcomnes/npm-run-all2#60]
I'm thinking it would be neat to be able to do:

```bash
run-p --workspaces check:*
```

Related issues:

* https://github.com/voxpelli/node-installed-check/issues/31
* https://github.com/dependency-check-team/dependency-check/issues/188
* 
Interesting modules:

* https://www.npmjs.com/package/@npmcli/map-workspaces
* https://www.npmjs.com/package/yarn-changed-workspaces

**Comment by bcomnes** [comment]
Makes sense now that npm supports workspaces.

**Comment by voxpelli** [comment]
When `--max-parallel` is set, should it try to spread out the executions across the workspaces or just assemble a long list of all tasks to be run and execute them from top to bottom?

The latter would be more a 1:1 between `run-s --workspaces` and `run-p --workspaces` but I kind of like the idea of balancing it across the different workspaces. Probably overengineering to do though, right?

**Comment by bcomnes** [comment]
Probably do what ever is simplest to implement and maintain.

**Comment by voxpelli** [comment]
I'm going to give it a stab now

**Comment by voxpelli** [comment]
Created a helper for reading all packages from a workspace, with similar filtering as npm itself has: https://github.com/voxpelli/read-workspaces

Started using it in `list-installed` and as such also in `installed-check` / `installed-check-core` now

**Comment by voxpelli** [comment]
RFC for providing `npm-run-all` style functionality within npm itself, with workspace support being one of the highlights: https://github.com/npm/rfcs/issues/190

Maybe we could try crowdfunding the development of this addition? Through eg. https://polar.sh/ or Open Collective?

**Comment by justinfagnani** [comment]
I've been advocating for [incorporating Wireit's script runner into npm](https://github.com/npm/rfcs/issues/706) - since I don't think that running scripts sequentially or in parallel is nearly sufficient.

I did hear from an npm team member after filing that issue that while they loved the idea, they didn't have the bandwidth to maintain such an addition, even if it were community contributed. I'm not sure if that situation still stands. A development since then is that Deno is adding a Wireit-inspired script runner as Deno Tasks. Maybe that moves forward the recognition that we need a more robust script running solution.

**Comment by bcomnes** [comment]
I haven't used wireit yet, but it looks interesting. Will give it a read through soon. run-s and run-p definitely have their limits, but also so does nearly every other tool in the npm ecosystem. Small modules almost never needed anything beyond what these two offered, but I also realize the tend these days is to try and operate entire companies out of workspaces, and so these simpler solutions fall over at that scale.

**Comment by voxpelli** [comment]
With more elaborate approaches the design choices also increase and it’s harder to align everyone on the same approach

It’s the classic small core vs big core argument – should it be solved within the core or does it belong in a pluralistic ecosystem

I think run-p and run-s style functionality could maybe belong in core but more than that: Probably not

**Comment by justinfagnani** [comment]
I don't think you need much scale before run-p and run-s fall over - you only need 4 scripts to create a diamond dependency where the deepest script should only run once.

A common setup might be a `test` script that depends on `test:unit` and `test:integration`, which both depend on `build`.

**Comment by voxpelli** [comment]
`run-s build test:*`

**Comment by justinfagnani** [comment]
@voxpelli the goal is to be able to not have to know how the script graph is structured in order to correctly run the scripts.

If something like this were included in npm, I would expect that `npm test` would just do the right thing _and_ that the config for `test` wouldn't have to list out its transitive dependencies.

**Comment by voxpelli** [comment]
@justinfagnani If you only have the 4 scripts then it works perfectly fine to do:

```json
"test": "run-s build test:*"
```

There is some point where it may be better for tasks to describe their own dependencies, but what percentage of npm projects are at such a complexity?

Many even find it enough to do so something like `eslint && mocha` or similar and use a single script for everything

**Comment by voxpelli** [comment]
A solution for now is to combine `npm query .workspace | jq -r '[.[].name] | join(" ")'` with the `{%}` placeholder:

```sh
    "dev-workspaces": "npm run dev:watch -w ",
    "dev:workspaces": "run-p \"dev-workspaces -- {%}\" -- $(npm query .workspace | jq -r '[.[].name] | join(\" \")')",
```

**Comment by bcomnes** [comment]
Another challenge here is pnpm has diverged on how it defines workspaces further from npm (it has its own file now for this). We could also just say this is a workspaces package.json field based feature.

**Comment by justinfagnani** [comment]
@bcomnes I don't thin that's too relevant for npm. pnpm could do many things that are different from and incompatible to npm.

**Comment by bcomnes** [comment]
Oh cool, didn't know about it.
